File: 222Project/src/graphing_package/Graph.java
package graphing_package;

public class Graph
{
   private static final int VERTEX_VISITED = 2;

   private Edge[] edgeSet;
   private Vertex[] vertexSet;
   public static int edgeIndex = 0;
   public static int vertexIndex = 0;

   /**
    * Creates a new Graph with known vertex and edge number
    * @param numberEdges The number of edges the graph will have
    * @param numbervertices The numnber of vertices the graph will have
    */
   public Graph(int numbervertices, int numberEdges)
   {
      edgeSet = new Edge [ numberEdges ];
      vertexSet = new Vertex [ numbervertices ];
   }
   
   /**
    * Adds an edge to the graph
    * <p>
    * Note: Creates a new edge object, does not use an already created edge
    * @param vertex1 The starting vertex of the edge
    * @param vertex2 The ending vertex for the edge
    * @param weight The weight of the edge
    */
   public void addEdge(Vertex vertex1, Vertex vertex2, int weight)
   {
      Edge newEdge = new Edge ( vertex1, vertex2, weight);
      edgeSet [ edgeIndex ] = newEdge;
      edgeIndex++;
      vertex1.addEdge( newEdge );
      vertex1.increaseDegree();
      vertex2.addEdge( newEdge );
      vertex2.increaseDegree();
   }
   
   /**
    * Adds a vertex to the graph
    * <p>
    * Note: Creates a new vertex object, does not use an already created vertex
    * @param name the name of the created vertex
    */
   public void addVertex (String name)
   {
      Vertex newVertex = new Vertex ( name );
      vertexSet [ vertexIndex ] = newVertex;
      vertexIndex++;
   }
   
   /**
    * Adds a vertex to the graph, with the automatic generated name
    * <p>
    * Note: Creates a new vertex object, does not use an already created vertex
    */
   public void addVertex()
   {
      Vertex newVertex = new Vertex ( "" );
      vertexSet [ vertexIndex ] = newVertex;
      vertexIndex++;
   }
   
   public int getNumVertices()
   {
      return vertexSet.length;
   }
   
   public int getNumEdges()
   {
      return edgeSet.length;
   }
   
   /**
    * Sorts all the edges of the graph's edge set
    * For use with the sorted edges algorithm
    * Note: uses selection sort
    */
   public void sortEdges()
   {
      int listIndex, innerIndex, smallestWeight,
         innerComparison, smallestWeightIndex;
      for ( listIndex = 0; listIndex < edgeSet.length - 1; listIndex++ )
      {
         smallestWeightIndex = listIndex;
         for ( innerIndex = listIndex + 1; innerIndex < edgeSet.length;
                  innerIndex++ )
         {
            smallestWeight = edgeSet[ smallestWeightIndex ].getWeight();
            innerComparison = edgeSet[ innerIndex ].getWeight();
            if ( innerComparison < smallestWeight )
            {
               smallestWeightIndex = innerIndex;
            }
         }
         swapElements( listIndex, smallestWeightIndex );
      }
   }
   
   
   private void swapElements( int oneIndex, int otherIndex )
   {
      Edge temp = edgeSet[ oneIndex ];
      
      edgeSet[ oneIndex ] =  edgeSet[ otherIndex ];
      
      edgeSet[ otherIndex ] =  temp;
      
   }
   
   /**
    * Runs the sorted edges Algorithm
    * @return The weight of the Set
    */
   public int runSortedEdges()
   {
      int[] verticesVisited = new int[ vertexSet.length ];//counts number of visits
      sortEdges();//sorts all the edges of the graph
      int edgeCounter = 0, edgeWeight = 0,
            vert1Num, vert2Num, vert1Visits, vert2Visits;
      Edge currEdge;
      boolean acceptableEdge, allVerticesVisited = false, graphWillBreak;
      while ( !allVerticesVisited )
      {
         acceptableEdge = false;
         while ( !acceptableEdge )
         {
            currEdge = edgeSet[ edgeCounter ];
            vert1Num = currEdge.getVertex1().getVertexNumber();
            vert1Visits = verticesVisited[ vert1Num ];
            vert2Num = currEdge.getVertex2().getVertexNumber();
            vert2Visits = verticesVisited[ vert2Num ];
            graphWillBreak = willEdgeBreakSort( verticesVisited, vert1Num, vert2Num  );
            if ( vert1Visits < VERTEX_VISITED && vert2Visits < VERTEX_VISITED
                  && !graphWillBreak )//checks if circuit will close early
               {
                  acceptableEdge = true;
                  edgeWeight += currEdge.getWeight();
                  verticesVisited[ vert1Num ]++;
                  verticesVisited[ vert2Num ]++;
               }
            edgeCounter++;
         }
         allVerticesVisited = areAllVerticesVisited( verticesVisited );
      }
      return edgeWeight;
   }
   
   private boolean willEdgeBreakSort( int[] verticesVisited, int vert1Num, int vert2Num  )
   {
      int vert1Val = verticesVisited[ vert1Num ];
      int vert2Val = verticesVisited[ vert2Num ];
      int vertexFinishedCounter = 0;
      int index;
      for ( index = 0; index < verticesVisited.length; index++ )
      {
         if ( verticesVisited[ index ] == 1 && index != vert1Num && index != vert2Num )
         {
            return false;
         }
         if ( verticesVisited[ index ] == VERTEX_VISITED )
         {
            vertexFinishedCounter++;
         }
      }
      if ( vertexFinishedCounter == vertexSet.length - 2 )
      {
         return false;
      }
      if ( vert1Val == 1 && vert2Val == 1 )
      {
         return true;
      }
      return false;
   }
   
   /**
    * Runs the nearest neighbor algorithm
    * @param startingVertex The vertex where the method starts
    * @return The total weight of the graph
    */
   public int runNearestNeighbor(Vertex startingVertex)//returns total weight
   {
      Vertex currVertex = startingVertex, otherVert;
      Edge shortEdge;
      int[] verticesVisited = new int [ vertexSet.length ];//showing vertices that have been visited
      int shortestEdgeTries, otherEdgeVertNum, edgeWeight = 0;
      boolean allVerticesVisited = false, acceptableEdge;
      
      int firstVertNum = startingVertex.getVertexNumber();
      verticesVisited[ firstVertNum ] += 2;
      
      while ( !allVerticesVisited )
      {
         acceptableEdge = false;
         shortestEdgeTries = 0;
         while ( !acceptableEdge )
         {
            shortEdge = currVertex.getShortestNeighbor( shortestEdgeTries );
            //current edge to be checked if acceptable
            otherVert = shortEdge.getOtherVertex( currVertex );
            otherEdgeVertNum = otherVert.getVertexNumber();
            acceptableEdge = verticesVisited [ otherEdgeVertNum ] != VERTEX_VISITED;
            if ( acceptableEdge )
            {
               verticesVisited[ otherEdgeVertNum ] += VERTEX_VISITED;//marks visited
               edgeWeight += shortEdge.getWeight();
               currVertex = otherVert;
               allVerticesVisited = areAllVerticesVisited( verticesVisited );
               if ( allVerticesVisited )
               {
                  edgeWeight +=
                     getToFirstVertex( currVertex, startingVertex ).getWeight();
               }
            }
            shortestEdgeTries++; 
         }

      }
      
      return edgeWeight;
   }
   
   
   private boolean areAllVerticesVisited( int[] verticesVisited )
   {
      int index;
      for ( index = 0; index < verticesVisited.length ; index++ )
      {
         if ( verticesVisited [ index ] < VERTEX_VISITED )
         {
            return false;
         }
      }
      return true;
   }
   
   private Edge getToFirstVertex( Vertex currVert, Vertex startingVertex )
   {
      Edge comparedEdge = new Edge( currVert, startingVertex, 0 );
      int index;
      for ( index = 0; index < edgeSet.length; index++ )
      {
         if ( comparedEdge.equals(edgeSet[index] ) )
         {
            return edgeSet[ index ];
         }
      }
      return null;
   }
   
   
   //TODO: should be in edge class
   /**
    * Takes in the name of an existing vertex, and returns object Vertex based
    * on that name
    * 
    * <p>
    * Note: will return null if the Vertex doesn't exist, but that should never 
    * occur here
    * Note: Two vertices of the same name cannot be in the same graph, 
    * else, the last vertex named that name will be found
    * 
    * @param name The name of the desired Vertex
    * @return The vertex that was requested
    */
   public Vertex findMeantVertex ( String name ) 
   {
      int index;
      String vertexName;
      for ( index = 0; index  < vertexSet.length; index++ )
      {
         vertexName = vertexSet[ index ].getName();
         if ( vertexName.equals( name ) )
         {
            return vertexSet[ index ];
         }
      }
      return null;
   }
   
   public Vertex findMeantVertex ( int vertNum )
   {
      return vertexSet[ vertNum ];
   }
   
   
   public void dumpVertices()
   {
      int index;
      for ( index = 0; index < vertexSet.length; index++ )
      {
         System.out.println( vertexSet[ index ].toString() );
      }
   }
   
   public void dumpEdges()
   {
      int index;
      for ( index = 0; index < edgeSet.length; index++ )
      {
         System.out.println( edgeSet[ index ].toString() );
      }
   }
}

File: PythonVersion/circuits/graph.py
import vertex
import edge

class Graph:
    #final variable
    VERTEX_VISITED = 2

    #class variable
    edge_index = 0
    vertex_index = 0

    def __init__( self ):
        self.edge_set = []
        self.vertex_set = []

    def add_edge( self, vertex1, vertex2, weight ):
        new_edge = edge.Edge( vertex1, vertex2, weight )
        self.edge_set.append( new_edge )
        vertex1.add_edge( new_edge )
        vertex1.increase_degree()
        vertex2.add_edge( new_edge )
        vertex2.increase_degree()

    def add_vertex( self, name ):
        new_vertex = vertex.Vertex( name )
        self.vertex_set.append( new_vertex )

    def get_num_vertices( self ):
        return len( self.vertex_set )

    def get_num_edges( self ):
        return len( self.edge_set )

    def sort_edges( self ):
        list_index = 0
        while list_index < len(self.edge_set) - 1:
            smallest_weight_index = list_index
            inner_index = list_index + 1
            while inner_index < len(self.edge_set):
                smallest_weight = self.edge_set[ smallest_weight_index ].get_weight()
                comparison = self.edge_set[ inner_index ].get_weight()
                if comparison < smallest_weight:
                    smallest_weight_index = inner_index
                inner_index += 1

            self.swap_elements( list_index, smallest_weight_index )    
            list_index += 1

    def swap_elements( self, one_index, other_index ):
        temp = self.edge_set[ one_index ]

        self.edge_set[ one_index ] = self.edge_set[ other_index ]

        self.edge_set[ other_index ] = temp

    def run_sorted_edges( self ):

        vertices_visited = [0] * len(self.vertex_set)
        self.sort_edges()
        edge_counter = 0
        edge_weight = 0
        all_vertices_visited = False

        while not all_vertices_visited:

            acceptable_edge = False

            while not acceptable_edge:
                
                curr_edge = self.edge_set[ edge_counter ]
                
                vert1_num = curr_edge.get_vertex1().get_vertex_number()
                vert1_visits = vertices_visited[ vert1_num ]
                vert2_num = curr_edge.get_vertex2().get_vertex_number()
                vert2_visits = vertices_visited[ vert2_num ]

                graph_will_break = self.is_edge_usable( vertices_visited, vert1_num, vert2_num )
                if vert1_visits < self.VERTEX_VISITED and vert2_visits < self.VERTEX_VISITED and not graph_will_break:
                    acceptable_edge = True
                    edge_weight += curr_edge.get_weight()
                    vertices_visited[ vert1_num ] += 1

                    vertices_visited[ vert2_num ] += 1

                edge_counter += 1

            all_vertices_visited = self.are_all_vertices_visited( vertices_visited )

        return edge_weight

    def is_edge_usable( self, vertices_visited, vert1_num, vert2_num ):

        vert1_val = vertices_visited[ vert1_num ]
        vert2_val = vertices_visited[ vert2_num ]
        vert_finished_count = 0
        index = 0

        while index < len(vertices_visited):
            if vertices_visited[ index ] == 1 and index != vert1_num and index != vert2_num:
                return False
            if vertices_visited[ index ] == self.VERTEX_VISITED:
                vert_finished_count += 1
            index += 1

        if vert_finished_count == len(self.vertex_set) - 2:
            return False
        if vert1_val == 1 and vert2_val == 1:
            return True
        return False

    def run_nearest_neighbor( self, starting_vertex ):

        curr_vert = starting_vertex;
        other_vert = None
        short_edge = None
        vertices_visited = [0] * len(self.vertex_set)
        edge_weight = 0
        all_vertices_visited = False

        vertices_visited[ starting_vertex.get_vertex_number() ] += 2

        while not all_vertices_visited:

            acceptable_edge = False
            shortest_edge_tries = 0
            
            while not acceptable_edge:
                
                short_edge = curr_vert.get_shortest_neighbor( shortest_edge_tries )
                other_vert = short_edge.get_other_vertex( curr_vert )
                other_vert_num = other_vert.get_vertex_number()

                acceptable_edge = vertices_visited[ other_vert_num ] != self.VERTEX_VISITED

                if acceptable_edge:

                   vertices_visited[ other_vert_num ] += self.VERTEX_VISITED
                   edge_weight += short_edge.get_weight()
                   curr_vert = other_vert
                   all_vertices_visited = self.are_all_vertices_visited( vertices_visited )

                   if all_vertices_visited:
                       edge_weight += self.get_starting_vertex( curr_vert, starting_vertex ).get_weight()

                shortest_edge_tries += 1
                
        return edge_weight


    def get_starting_vertex( self, curr_vert, starting_vertex ):

        compared_edge = edge.Edge( curr_vert, starting_vertex, 0 )
        index = 0
        while index < len(self.edge_set):
            if compared_edge.equals( self.edge_set[ index ] ):
                return self.edge_set[ index ]
            index += 1
        return None

        
    def are_all_vertices_visited( self, vertices_visited ):

        index = 0
        while index < len(vertices_visited):

            if vertices_visited[ index ] < self.VERTEX_VISITED:
                return False
            index += 1
        return True

    def find_meant_vertex( self, name ):

        if type( name ) is int:
            return self.vertex_set[ name ]
        
        index = 0

        while index < len(self.vertex_set):

            vertex_name = self.vertex_set[ index ].get_name()
            if vertex_name == name:
                return self.vertex_set[ index ]
            index += 1

        return None

    def dump_vertices( self ):

        print ( self.vertex_set )

    def dump_edges( self ):

        print ( self.edge_set )
